src/Products/PluginIndexes/KeywordIndex/KeywordIndex.py

```diff
@@ -58,59 +72,120 @@ def _index_object(self, documentId, obj, threshold=None, attr=''):
         # we'll do so.
 
         newKeywords = self._get_object_keywords(obj, attr)
+        oldKeywords = self._unindex.get(documentId, _marker)
 
-        oldKeywords = self._unindex.get(documentId, None)
-
-        if oldKeywords is None:
+        if oldKeywords is _marker:
             # we've got a new document, let's not futz around.
-            try:
+            if newKeywords in (missing, empty):
+                self.insertSpecialIndexEntry(newKeywords, documentId)
+            else:
+                keys = list()
                 for kw in newKeywords:
-                    self.insertForwardIndexEntry(kw, documentId)
-                if newKeywords:
-                    self._unindex[documentId] = list(newKeywords)
-            except TypeError:
-                return 0
+                    try:
+                        self.insertForwardIndexEntry(kw, documentId)
+                        keys.append(kw)
+                    except TypeError:
+                        # key is not valid for this Btree so we have to
+                        # roll back insertForwardIndexEntry
+                        LOG.error('%(context)s: Unable to insert forward '
+                                  'index entry for document with id '
+                                  '%(doc_id)s and keyword %(kw)r '
+                                  'for index %{index}r.', dict(
+                                      context=self.__class__.__name__,
+                                      kw=kw,
+                                      doc_id=documentId,
+                                      index=self.id))
+
+                        self.unindex_objectKeywords(documentId, keys)
+                        return 0
+
+                newKeywords = OOSet(newKeywords)
+
+            self._unindex[documentId] = newKeywords
```

Comment:
The `_unindex` update could be done together with the "update existing index info" case.

Reply:
Okay